main/src/library/Fixpoint3/Interpreter.flix

```diff
@@ -0,0 +1,735 @@
+/*
+ * Copyright 2021 Benjamin Dahse
+ * Copyright 2025 Casper Dalgaard Nielsen
+ *                Adam Yasser Tallouzi
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+mod Fixpoint3.Interpreter {
+    use Fixpoint3.Ast.ExecutableRam.{BoolExp, RamProgram, RamStmt, RamTerm, RelOp, WriteTuple}
+    use Fixpoint3.Ast.Ram.{arityOfNonLat, Predicates, RamId, RelSym, Search, toDenotation, toId}
+    use Fixpoint3.Ast.Shared.{BoxedDenotation => Denotation}
+    use Fixpoint3.Ast.Shared.Denotation.{Latticenal, Relational}
+    use Fixpoint3.Boxed
+    use Fixpoint3.Boxing.{boxWith, unboxWith}
+    use Fixpoint3.BoxingType.{Boxing, RamIdToPos}
+    use Fixpoint3.Counter
+    use Fixpoint3.Options.{usedArity, parLevel}
+    use Fixpoint3.Predicate.allFullRelSyms
+    use Fixpoint3.Util.getOrCrash
+
+    ///
+    /// A tuple/fact.
+    ///
+    type alias Tuple = Vector[Int64]
+
+    ///
+    /// The number of times the interpreter can recursively split a `RelOp.Search` into 
+    /// smaller workloads for threads running in parallel.
+    ///
+    /// Consider the program:
+    /// ```
+    /// search a ∈ A do
+    ///     search b ∈ B do
+    ///         search c ∈ C do
+    ///             ...
+    ///         end
+    ///     end
+    /// end
+    /// ```
+    /// Spawning `k` threads in every layer with a `RelOp.Search`s undesirable, since this 
+    /// leads to `k^d` threads for a nesting of depth `d`. Therefore, the interpreter is 
+    /// limited to only spawn threads for the `MaxParallelNesting` highest `search` `RelOp`s.
+    ///
+    /// For the example, if we allow a maximum of 2 parallel nestings, we would not spawn
+    /// threads when evaluating `search c ∈ C do`.
+    ///
+    type alias MaxParallelNesting = Int32
```

Comment:
This seems like an odd type alias to me. When its used, it's also for a number that is counting down, so it'd be CurrentParallelNesting if anything. Imo This should be removed and the documentation should be put onto Options.parLevel